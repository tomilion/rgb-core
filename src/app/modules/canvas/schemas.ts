export interface CreateCanvasPayload {
    canvasId: number;
    costPerPixel: bigint;
    startBlockHeight: bigint;
    endBlockHeight: bigint;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: Buffer;
    maxPixelsPerTransaction: number;
}

export interface ChangeCanvasPayload {
    canvasId: number;
    costPerPixel?: bigint | null;
    startBlockHeight?: bigint | null;
    endBlockHeight?: bigint | null;
    width?: number | null;
    height?: number | null;
    timeBetweenDraws?: number | null;
    colourPalette?: Buffer | null;
    maxPixelsPerTransaction?: number | null;
}

export enum CanvasState {
    PENDING = 0,
    ACTIVE = 1,
    COMPLETE = 2,
}

export interface CanvasPayload {
    ownerId: Buffer;
    costPerPixel: bigint;
    startBlockHeight: bigint;
    endBlockHeight: bigint;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: Buffer;
    maxPixelsPerTransaction: number;
    state: number;
}

export const canvasSchema = {
    $id: "canvas/canvas",
    type: "object",
    required: ["ownerId", "costPerPixel", "startBlockHeight", "endBlockHeight", "width", "height", "timeBetweenDraws", "colourPalette", "maxPixelsPerTransaction"],
    properties: {
        ownerId: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
        costPerPixel: { fieldNumber: 2, dataType: "uint64" },
        startBlockHeight: { fieldNumber: 3, dataType: "uint64" },
        endBlockHeight: { fieldNumber: 4, dataType: "uint64" },
        width: { fieldNumber: 5, dataType: "uint32" },
        height: { fieldNumber: 6, dataType: "uint32" },
        timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
        colourPalette: { fieldNumber: 8, dataType: "bytes", minLength: 48, maxLength: 48 },
        maxPixelsPerTransaction: { fieldNumber: 9, dataType: "uint32" },
        state: { fieldNumber: 10, dataType: "uint32", default: CanvasState.PENDING },
    },
};

export interface AccountPayload {
    lastBlockHeight: number;
}

export const accountSchema = {
    $id: "canvas/account",
    type: "object",
    required: ["lastBlockHeight"],
    properties: {
        lastBlockHeight: { fieldNumber: 1, dataType: "uint32" },
    },
};

export interface PendingPayload {
    canvasIds: number[];
}

export const pendingSchema = {
    $id: "canvas/pending",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface ActivePayload {
    canvasIds: number[];
}

export const activeSchema = {
    $id: "canvas/active",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface CompletePayload {
    canvasIds: number[];
}

export const completeSchema = {
    $id: "canvas/complete",
    type: "object",
    required: ["canvasIds"],
    properties: {
        canvasIds: { fieldNumber: 1, type: "array", items: { dataType: 'uint32' } },
    },
};

export interface AddressPayload {
    address: Buffer;
}

export const addressSchema = {
    $id: "canvas/address",
    type: "object",
    required: ["address"],
    properties: {
        address: { fieldNumber: 1, dataType: "bytes", minLength: 20, maxLength: 20 },
    },
};

export interface DrawPixelPayload {
    canvasId: number;
    coords: Buffer;
    colours: Buffer;
}

export interface DrawPixelPayloadJSON {
    canvasId: number;
    coords: string;
    colours: string;
}

export const drawPixelSchema = {
    $id: "canvas/drawPixel/asset",
    title: "DrawPixelAsset transaction asset for Canvas module",
    type: "object",
    required: ["canvasId", "coords", "colours"],
    properties: {
        canvasId: { fieldNumber: 1, dataType: "uint32" },
        coords: { fieldNumber: 2, dataType: "bytes", minLength: 3, maxLength: 30000 },
        colours: { fieldNumber: 3, dataType: "bytes", minLength: 1, maxLength: 5000 },
    },
};

export interface CanvasAccount {
    canvas: { accountType: string }
}

export enum AccountType {
    Admin = "ADMIN",
    Wallet = "WALLET",
    Default = "DEFAULT",
}

export interface CanvasResponse {
    ownerId: string;
    costPerPixel: number;
    startBlockHeight: number;
    endBlockHeight: number;
    width: number;
    height: number;
    timeBetweenDraws: number;
    colourPalette: string;
    maxPixelsPerTransaction: number;
    state: number;
}

export type CanvasId = {
    canvasId: number;
};

export type PixelChangeSubmitted = {
    address: string;
    transactionId: string;
    pixel: DrawPixelPayload;
};

export type PixelChangeCommitted = {
    address: string;
    transactionId: string;
    blockHeight: number;
    pixel: DrawPixelPayload;
};
