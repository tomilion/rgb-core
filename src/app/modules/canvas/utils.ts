export const serialiseCanvasId = (canvasId: number) => `canvas-${canvasId}`;

export const serialisePixelId = (canvasId: number, xCoord: number, yCoord: number) =>
    `${serialiseCanvasId(canvasId)}-pixel-${xCoord}-${yCoord}`;

export const serialiseAccountId = (canvasId: number, address: string) => `${serialiseCanvasId(canvasId)}-account-${address}`;
