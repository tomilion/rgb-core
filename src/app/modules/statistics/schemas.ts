export interface TotalTransactionsPayload {
    totalTransactions: number;
}

export const totalTransactionsSchema = {
    $id: "statistics/totalTransactions",
    type: "object",
    required: ["totalTransactions"],
    properties: {
        totalTransactions: { fieldNumber: 1, dataType: "uint32" },
    },
};
