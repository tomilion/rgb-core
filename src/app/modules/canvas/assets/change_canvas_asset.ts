import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseCanvasId } from "../utils";
import { CanvasPayload, canvasSchema, CanvasState, ChangeCanvasPayload } from "../schemas";

export class ChangeCanvasAsset extends BaseAsset<ChangeCanvasPayload> {
    public static readonly ASSET_ID = 1;

    public name = "changeCanvas";
    public id = ChangeCanvasAsset.ASSET_ID;
    public schema = {
        $id: "canvas/changeCanvas-asset",
        title: "ChangeCanvasAsset transaction asset for canvas module",
        type: "object",
        required: ["canvasId"],
        properties: {
            canvasId: { fieldNumber: 1, dataType: "uint32" },
            costPerPixel: { fieldNumber: 2, dataType: "uint64" },
            startBlockHeight: { fieldNumber: 3, dataType: "uint64" },
            endBlockHeight: { fieldNumber: 4, dataType: "uint64" },
            width: { fieldNumber: 5, dataType: "uint32" },
            height: { fieldNumber: 6, dataType: "uint32" },
            timeBetweenDraws: { fieldNumber: 7, dataType: "uint32" },
            colourPalette: { fieldNumber: 8, dataType: "bytes", minLength: 48, maxLength: 48 },
            maxPixelsPerTransaction: { fieldNumber: 9, dataType: "uint32" },
            label: { fieldNumber: 10, dataType: "string", minLength: 0, maxLength: 64 },
        },
    };

    public validate(context: ValidateAssetContext<ChangeCanvasPayload>): void {
        const { asset, header } = context;

        if (asset.width !== undefined && asset.width !== null && (asset.width < 1 || asset.width > 10000))
        {
            throw new Error("Width invalid");
        }

        if (asset.height !== undefined && asset.height !== null && (asset.height < 1 || asset.height > 10000))
        {
            throw new Error("Height invalid");
        }

        if (asset.costPerPixel !== undefined && asset.costPerPixel !== null && asset.costPerPixel < 0)
        {
            throw new Error("Cost per pixel invalid");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.startBlockHeight !== undefined && asset.startBlockHeight !== null && asset.startBlockHeight < header.height)
        {
            throw new Error("Start block height cannot be in the past");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.endBlockHeight !== undefined && asset.endBlockHeight !== null && asset.endBlockHeight < header.height)
        {
            throw new Error("End block height cannot be in the past");
        }

        // Checking undefined/null inline because storing condition in variable is too complex for linter ...
        if (asset.startBlockHeight !== undefined &&
            asset.startBlockHeight !== null &&
            asset.endBlockHeight !== undefined &&
            asset.endBlockHeight !== null &&
            asset.endBlockHeight < asset.startBlockHeight)
        {
            throw new Error("End block height must be greater than start block height");
        }

        if (asset.maxPixelsPerTransaction !== undefined && asset.maxPixelsPerTransaction !== null && (asset.maxPixelsPerTransaction < 1 || asset.maxPixelsPerTransaction > 10000))
        {
            throw new Error("Max pixels per transaction invalid");
        }

        if (asset.colourPalette !== undefined && asset.colourPalette !== null && asset.colourPalette.length !== 48)
        {
            throw new Error("Colour palette invalid");
        }
    }

    public async apply(context: ApplyAssetContext<ChangeCanvasPayload>): Promise<void> {
        const { asset, stateStore, transaction } = context;
        const canvasId = serialiseCanvasId(asset.canvasId);
        const currentCanvas = await stateStore.chain.get(canvasId);

        if (currentCanvas === undefined)
        {
            throw new Error("Canvas does not exist");
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, currentCanvas);

        if (canvas.ownerId.toString("hex") !== transaction.senderAddress.toString("hex"))
        {
            throw new Error("User invalid");
        }

        if (canvas.state === CanvasState.COMPLETE)
        {
            throw new Error("Canvas already completed");
        }

        if (canvas.state === CanvasState.ACTIVE)
        {
            if (asset.startBlockHeight !== null && asset.startBlockHeight !== undefined)
            {
                throw new Error("Cannot modify start block height on active canvas");
            }

            if (asset.endBlockHeight !== null && asset.endBlockHeight !== undefined)
            {
                throw new Error("Cannot modify end block height on active canvas");
            }
        }

        canvas.width = asset.width ?? canvas.width;
        canvas.height = asset.height ?? canvas.height;
        canvas.costPerPixel = asset.costPerPixel ?? canvas.costPerPixel;
        canvas.startBlockHeight = asset.startBlockHeight ?? canvas.startBlockHeight;
        canvas.endBlockHeight = asset.endBlockHeight ?? canvas.endBlockHeight;
        canvas.timeBetweenDraws = asset.timeBetweenDraws ?? canvas.timeBetweenDraws;
        canvas.colourPalette = asset.colourPalette ?? canvas.colourPalette;
        canvas.maxPixelsPerTransaction = asset.maxPixelsPerTransaction ?? canvas.maxPixelsPerTransaction;
        canvas.label = asset.label ?? canvas.label;

        if (canvas.startBlockHeight > canvas.endBlockHeight)
        {
            throw new Error("End block height must be greater than start block height");
        }

        await stateStore.chain.set(canvasId, codec.encode(canvasSchema, canvas));
    }
}
