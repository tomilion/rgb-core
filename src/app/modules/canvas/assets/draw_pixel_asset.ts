import { ApplyAssetContext, BaseAsset, codec, ValidateAssetContext } from "lisk-sdk";
import { serialiseAccountId, serialiseCanvasId } from "../utils";
import { AccountPayload, accountSchema, CanvasPayload, canvasSchema, CanvasState, DrawPixelPayload, drawPixelSchema } from "../schemas";

export class DrawPixelAsset extends BaseAsset<DrawPixelPayload> {
    public static readonly ASSET_ID = 5;

    public name = "drawPixel";
    public id = DrawPixelAsset.ASSET_ID;
    public schema = drawPixelSchema;

    public validate(context: ValidateAssetContext<DrawPixelPayload>): void {
        const { asset } = context;

        if (asset.colours.length < 1)
        {
            throw new Error("Requires at least 1 colour");
        }

        // 24 bit coordinates (supports max canvas size of approx 4k * 4k)
        if ((asset.coords.length % 3) > 0)
        {
            throw new Error("Coords invalid");
        }

        if (Math.ceil(asset.coords.length / 3 / 2) !== asset.colours.length)
        {
            throw new Error("Number of coords does not match number of colours");
        }
    }

    public async apply(context: ApplyAssetContext<DrawPixelPayload>): Promise<void> {
        const { asset, transaction, stateStore, reducerHandler } = context;
        const canvasId = serialiseCanvasId(asset.canvasId);
        const canvasBuffer = await stateStore.chain.get(canvasId);

        if (canvasBuffer === undefined)
        {
            throw new Error("Canvas does not exist");
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);

        if (canvas.state !== CanvasState.ACTIVE)
        {
            throw new Error("Canvas not active");
        }

        if ((asset.coords.length / 3) > canvas.maxPixelsPerTransaction)
        {
            throw new Error("Too many pixels");
        }

        for (let i = 0; i < asset.coords.length; i += 3)
        {
            // 24 bit coordinate packed with little endianness
            const coord = asset.coords[i] | (asset.coords[i + 1] << 8) | (asset.coords[i + 2] << 16);

            // Coords encoded as numbered position in canvas eg. (1, 1) in a 1000 x 1000 canvas would be encoded as (1001)
            // where 0 -> 999 is y = 0 and 1000 -> 1999 is y = 1 (both cases x = {0 -> 999})
            if (coord >= (canvas.width * canvas.height))
            {
                throw new Error("Coords invalid");
            }
        }

        // Check user has waited for draw timeout before drawing again
        const accountId = serialiseAccountId(asset.canvasId, transaction.senderAddress.toString("hex"));
        const accountBuffer = await stateStore.chain.get(accountId);
        const account = (accountBuffer !== undefined) ? codec.decode<AccountPayload>(accountSchema, accountBuffer) : { lastBlockHeight: 0 };
        const lastBlock = stateStore.chain.lastBlockHeaders[stateStore.chain.lastBlockHeaders.length - 1];

        if ((account.lastBlockHeight + canvas.timeBetweenDraws) > (lastBlock.height + 1))
        {
            throw new Error("Too many draws");
        }

        const walletAddress = await reducerHandler.invoke<Buffer>("canvas:getWalletAddress");

        await reducerHandler.invoke("token:debit", {
            address: transaction.senderAddress,
            amount: canvas.costPerPixel,
        });

        await reducerHandler.invoke("token:credit", {
            address: walletAddress,
            amount: canvas.costPerPixel,
        });

        // Set last draw time
        await stateStore.chain.set(accountId, codec.encode(accountSchema, { lastBlockHeight: lastBlock.height + 1 }));
    }
}
