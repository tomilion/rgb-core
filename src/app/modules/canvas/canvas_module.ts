import { AfterBlockApplyContext, AfterGenesisBlockApplyContext, BaseModule, BeforeBlockApplyContext, codec, TransactionApplyContext } from "lisk-sdk";
import { strict as assert } from "assert";
import { ChangeCanvasAsset } from "./assets/change_canvas_asset";
import { CreateCanvasAsset } from "./assets/create_canvas_asset";
import { DrawPixelAsset, DrawPixelPayload, drawPixelSchema } from "./assets/draw_pixel_asset";
import { serialiseCanvasId } from "./utils";
import { AccountType, ActivePayload, activeSchema, AddressPayload, addressSchema, CanvasAccount, CanvasPayload, CanvasResponse, canvasSchema, CanvasState, CompletePayload, completeSchema, PendingPayload, pendingSchema } from "./schemas";

export class CanvasModule extends BaseModule {
    public static readonly MODULE_ID = 1000;

    public name = "canvas";
    public id = CanvasModule.MODULE_ID;
    public accountSchema = {
        type: "object",
        properties: {
            accountType: {
                fieldNumber: 1,
                dataType: "string",
                maxLength: 64,
            },
        },
        default: {
            accountType: AccountType.Default.toString(),
        },
    };
    public actions = {
        getCanvas: async (params) => this.getCanvas(params),
        getPendingCanvases: async () => this.getPendingCanvases(),
        getActiveCanvases: async () => this.getActiveCanvases(),
        getCompleteCanvases: async () => this.getCompleteCanvases(),
    };
    public reducers = {
        getWalletAddress: async () => this.getWalletAddress(),
        getAdminAddress: async () => this.getAdminAddress(),
    };
    public transactionAssets = [
        new CreateCanvasAsset(),
        new ChangeCanvasAsset(),
        new DrawPixelAsset(),
    ];
    public events = ["started", "completed", "pixelChangeSubmitted", "pixelChangeCommitted"];

    // Lifecycle hooks
    public async beforeBlockApply(_input: BeforeBlockApplyContext): Promise<void> {
        // Get any data from stateStore using block info, below is an example getting a generator
        // const generatorAddress = getAddressFromPublicKey(_input.block.header.generatorPublicKey);
        // const generator = await _input.stateStore.account.get<TokenAccount>(generatorAddress);
    }

    public async afterBlockApply(context: AfterBlockApplyContext): Promise<void> {
        const { block, stateStore } = context;
        const pendingToCommit: number[] = [];
        const activeToCommit: number[] = [];

        const pendingBuffer = await stateStore.chain.get("canvas:pending");
        const pending = (pendingBuffer !== undefined) ? codec.decode<PendingPayload>(pendingSchema, pendingBuffer) : { canvasIds: [] };

        const activeBuffer = await stateStore.chain.get("canvas:active");
        const active = (activeBuffer !== undefined) ? codec.decode<ActivePayload>(activeSchema, activeBuffer) : { canvasIds: [] };

        const completeBuffer = await stateStore.chain.get("canvas:complete");
        const complete = (completeBuffer !== undefined) ? codec.decode<CompletePayload>(completeSchema, completeBuffer) : { canvasIds: [] };

        for (const canvasId of pending.canvasIds)
        {
            const canvasBuffer = await stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.PENDING);

            // Automatically starting canvas after passing start time
            if (canvas.startBlockHeight <= block.header.height)
            {
                canvas.state = CanvasState.ACTIVE;
                await stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                activeToCommit.push(canvasId);
                this._channel.publish("canvas:started", { canvasId });
                continue;
            }

            pendingToCommit.push(canvasId);
        }

        for (const transaction of block.payload)
        {
            if (transaction.moduleID === CanvasModule.MODULE_ID &&
                transaction.assetID === DrawPixelAsset.ASSET_ID)
            {
                const pixel = codec.decode<DrawPixelPayload>(drawPixelSchema, transaction.asset);
                this._channel.publish("canvas:pixelChangeCommitted", {
                    address: transaction.senderAddress.toString("hex"),
                    transactionId: transaction.id.toString("hex"),
                    blockHeight: block.header.height,
                    pixel: pixel,
                });
            }
        }

        for (const canvasId of active.canvasIds)
        {
            const canvasBuffer = await stateStore.chain.get(serialiseCanvasId(canvasId));
            assert(canvasBuffer !== undefined);
            const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
            assert(canvas.state === CanvasState.ACTIVE);

            // Automatically ending canvas after passing end time
            if (canvas.endBlockHeight <= block.header.height)
            {
                canvas.state = CanvasState.COMPLETE;
                await stateStore.chain.set(serialiseCanvasId(canvasId), codec.encode(canvasSchema, canvas));
                complete.canvasIds.push(canvasId);
                this._channel.publish("canvas:completed", { canvasId });
                continue;
            }

            activeToCommit.push(canvasId);
        }

        await stateStore.chain.set("canvas:pending", codec.encode(pendingSchema, { canvasIds: pendingToCommit }));
        await stateStore.chain.set("canvas:active", codec.encode(activeSchema, { canvasIds: activeToCommit }));
        await stateStore.chain.set("canvas:complete", codec.encode(completeSchema, complete));
    }

    public async beforeTransactionApply(_input: TransactionApplyContext): Promise<void> {
        // Get any data from stateStore using transaction info, below is an example
        // const sender = await _input.stateStore.account.getOrDefault<TokenAccount>(_input.transaction.senderAddress);
    }

    public async afterTransactionApply(_input: TransactionApplyContext): Promise<void> {
        if (_input.transaction.moduleID === CanvasModule.MODULE_ID &&
            _input.transaction.assetID === DrawPixelAsset.ASSET_ID)
        {
            const pixel = codec.decode<DrawPixelPayload>(drawPixelSchema, _input.transaction.asset);
            this._channel.publish("canvas:pixelChangeSubmitted", {
                address: _input.transaction.senderAddress.toString("hex"),
                transactionId: _input.transaction.id.toString("hex"),
                pixel: pixel,
            });
        }
    }

    public async afterGenesisBlockApply(_input: AfterGenesisBlockApplyContext<CanvasAccount>): Promise<void> {
        const accounts = _input.genesisBlock.header.asset.accounts;
        assert(accounts.length > 0);

        for (const account of accounts)
        {
            if (account.canvas.accountType == AccountType.Admin.toString())
            {
                await _input.stateStore.chain.set("canvas:adminAddress", codec.encode(addressSchema, { address: account.address }));
            }

            if (account.canvas.accountType == AccountType.Wallet.toString())
            {
                await _input.stateStore.chain.set("canvas:walletAddress", codec.encode(addressSchema, { address: account.address }));
            }
        }
    }

    private async getCanvas(params: { canvasId: string }): Promise<CanvasResponse|null> {
        const canvasId = serialiseCanvasId(Number(params.canvasId));
        const canvasBuffer = await this._dataAccess.getChainState(canvasId);

        if (canvasBuffer === undefined)
        {
            return null;
        }

        const canvas = codec.decode<CanvasPayload>(canvasSchema, canvasBuffer);
        return {
            ...canvas,
            ownerId: canvas.ownerId.toString("hex"),
            costPerPixel: Number(canvas.costPerPixel),
            startBlockHeight: Number(canvas.startBlockHeight),
            endBlockHeight: Number(canvas.endBlockHeight),
        };
    }

    private async getPendingCanvases(): Promise<PendingPayload> {
        const pendingBuffer = await this._dataAccess.getChainState("canvas:pending");

        if (pendingBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<PendingPayload>(pendingSchema, pendingBuffer);
    }

    private async getActiveCanvases(): Promise<ActivePayload> {
        const activeBuffer = await this._dataAccess.getChainState("canvas:active");

        if (activeBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<ActivePayload>(activeSchema, activeBuffer);
    }

    private async getCompleteCanvases(): Promise<CompletePayload> {
        const completeBuffer = await this._dataAccess.getChainState("canvas:complete");

        if (completeBuffer === undefined)
        {
            return { canvasIds: [] };
        }

        return codec.decode<CompletePayload>(completeSchema, completeBuffer);
    }

    private async getWalletAddress(): Promise<Buffer|null> {
        const walletBuffer: Buffer|undefined = await this._dataAccess.getChainState("canvas:walletAddress");

        if (walletBuffer === undefined)
        {
            return null;
        }

        const wallet = codec.decode<AddressPayload>(addressSchema, walletBuffer);
        return wallet.address;
    }

    private async getAdminAddress(): Promise<Buffer|null> {
        const adminBuffer: Buffer|undefined = await this._dataAccess.getChainState("canvas:adminAddress");

        if (adminBuffer === undefined)
        {
            return null;
        }

        const admin = codec.decode<AddressPayload>(addressSchema, adminBuffer);
        return admin.address;
    }
}
