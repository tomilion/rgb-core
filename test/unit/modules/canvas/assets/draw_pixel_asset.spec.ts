import { ReducerHandler } from "lisk-framework/dist-node/types";
import { codec, testing } from "lisk-sdk";
import { Account } from "@liskhq/lisk-chain/dist-node/types";
import { when } from "jest-when";
import { DrawPixelAsset } from "../../../../../src/app/modules/canvas/assets/draw_pixel_asset";
import { CanvasModule } from "../../../../../src/app/modules/canvas/canvas_module";
import { accountSchema, CanvasPayload, canvasSchema, CanvasState, DrawPixelPayload } from "../../../../../src/app/modules/canvas/schemas";
import { numberBetween, randomCanvas, randomCoordinate, randomDrawPixel } from "../../../../utils/random_generator";

describe("DrawPixelAsset", () => {
    let mockAsset: DrawPixelPayload;
    let testClass: DrawPixelAsset;

    beforeEach(() => {
        mockAsset = randomDrawPixel();
        testClass = new DrawPixelAsset();
    });

    describe("constructor", () => {
        it("should have valid id", () => {
            expect(testClass.id).toEqual(5);
        });

        it("should have valid name", () => {
            expect(testClass.name).toEqual("drawPixel");
        });

        it("should have valid schema", () => {
            expect(testClass.schema).toMatchSnapshot();
        });
    });

    describe("validate", () => {
        it("should pass with valid args", () => {
            const context = testing.createValidateAssetContext({
                asset: mockAsset,
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            testClass.validate(context);
        });

        it("should throw no colours provided", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, colours: new Uint8Array(0) },
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Requires at least 1 colour");
        });

        it("should throw too many colours provided", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, colours: new Uint8Array(51) },
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Exceeded maximum number of colours");
        });

        it("should throw coords do not fit multiple of 24 bit", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, coords: new Uint8Array(4) },
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Coords invalid");
        });

        it("should throw not enough coords for colours", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, coords: new Uint8Array(9), colours: new Uint8Array(3) },
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Number of coords does not match number of colours");
        });

        it("should throw not enough colours for coords", () => {
            const context = testing.createValidateAssetContext({
                asset: { ...mockAsset, coords: new Uint8Array(9), colours: new Uint8Array(1) },
                transaction: { senderAddress: Buffer.alloc(0) } as any,
            });
            expect(() => testClass.validate(context)).toThrow("Number of coords does not match number of colours");
        });
    });

    describe("apply", () => {
        let account: Account;
        let chain: any;
        let canvas: CanvasPayload;
        let reducerHandler: ReducerHandler;

        beforeEach(() => {
            account = testing.fixtures.createDefaultAccount([CanvasModule]);
            chain = {};
            canvas = randomCanvas({
                width: 1000,
                height: 1000,
                timeBetweenDraws: numberBetween(1, 900),
            });
            reducerHandler = testing.mocks.reducerHandlerMock;
        });

        it("should update existing canvas account in state store", async () => {
            const height = numberBetween(canvas.timeBetweenDraws, canvas.timeBetweenDraws + 100000);
            const lastBlockHeaders = [{ height: height - 1 }, { height: height }];

            mockAsset.coords = new Uint8Array(randomCoordinate(canvas.width, canvas.height));
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, canvas);
            chain[`canvas-${mockAsset.canvasId}-account-${account.address.toString("hex")}`] = codec.encode(accountSchema, {
                lastBlockHeight: height - canvas.timeBetweenDraws
            });

            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain,
                lastBlockHeaders,
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });
            const setMock = jest.spyOn(stateStore.chain, "set");
            const invokeMock = jest.spyOn(reducerHandler, "invoke");

            when(invokeMock).calledWith("canvas:getLastBlockHeight").mockResolvedValue(height);

            await testClass.apply(context);

            expect(setMock).toHaveBeenCalledWith(
                `canvas-${mockAsset.canvasId}-account-${account.address.toString("hex")}`,
                codec.encode(accountSchema, { lastBlockHeight: height + 1 })
            );
        });

        it("should allow multiples draws to same block with timeBetweenDraws of zero", async () => {
            const height = numberBetween(100, 100000);
            const lastBlockHeaders = [{ height: height - 1 }, { height: height }];
            canvas.timeBetweenDraws = 0;

            mockAsset.coords = new Uint8Array(randomCoordinate(canvas.width, canvas.height));
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, canvas);
            chain[`canvas-${mockAsset.canvasId}-account-${account.address.toString("hex")}`] = codec.encode(accountSchema, {
                lastBlockHeight: height + 1
            });

            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain,
                lastBlockHeaders,
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });
            const setMock = jest.spyOn(stateStore.chain, "set");
            const invokeMock = jest.spyOn(reducerHandler, "invoke");

            when(invokeMock).calledWith("canvas:getLastBlockHeight").mockResolvedValue(height);

            await testClass.apply(context);

            expect(setMock).toHaveBeenCalledWith(
                `canvas-${mockAsset.canvasId}-account-${account.address.toString("hex")}`,
                codec.encode(accountSchema, { lastBlockHeight: height + 1 })
            );
        });

        it("should throw canvas does not exist", async () => {
            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });

            await expect(testClass.apply(context)).rejects.toThrow("Canvas does not exist");
        });

        it("should throw canvas not active", async () => {
            canvas.state = CanvasState.COMPLETE;
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, canvas);

            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });

            await expect(testClass.apply(context)).rejects.toThrow("Canvas not active");
        });

        it("should throw coords out of bounds", async () => {
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, canvas);
            const coord = canvas.width * canvas.height;
            mockAsset.coords = new Uint8Array([coord & 0xFF, (coord >> 8) & 0xFF, (coord >> 16) & 0xFF]);

            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });

            await expect(testClass.apply(context)).rejects.toThrow("Coords invalid");
        });

        it("should throw user not waited for draw timeout", async () => {
            const height = numberBetween(canvas.timeBetweenDraws, canvas.timeBetweenDraws + 100000);
            const lastBlockHeaders = [{ height: height - 1 }, { height: height }];

            mockAsset.coords = new Uint8Array(randomCoordinate(canvas.width, canvas.height));
            chain[`canvas-${mockAsset.canvasId}`] = codec.encode(canvasSchema, canvas);
            chain[`canvas-${mockAsset.canvasId}-account-${account.address.toString("hex")}`] = codec.encode(accountSchema, {
                lastBlockHeight: height - canvas.timeBetweenDraws + 2
            });

            const stateStore = new testing.mocks.StateStoreMock({
                accounts: [account],
                chain,
                lastBlockHeaders
            });
            const context = testing.createApplyAssetContext({
                transaction: { senderAddress: account.address, nonce: BigInt(1) } as any,
                asset: mockAsset,
                reducerHandler,
                stateStore
            });
            const invokeMock = jest.spyOn(reducerHandler, "invoke");

            when(invokeMock).calledWith("canvas:getLastBlockHeight").mockResolvedValue(height);

            await expect(testClass.apply(context)).rejects.toThrow("Too many draws");
        });
    });
});
