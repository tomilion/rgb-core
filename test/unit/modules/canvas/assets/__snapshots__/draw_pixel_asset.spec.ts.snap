// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawPixelAsset constructor should have valid schema 1`] = `
Object {
  "$id": "canvas/drawPixel/asset",
  "properties": Object {
    "canvasId": Object {
      "dataType": "uint32",
      "fieldNumber": 1,
    },
    "colours": Object {
      "dataType": "bytes",
      "fieldNumber": 3,
      "maxLength": 5000,
      "minLength": 1,
    },
    "coords": Object {
      "dataType": "bytes",
      "fieldNumber": 2,
      "maxLength": 30000,
      "minLength": 3,
    },
  },
  "required": Array [
    "canvasId",
    "coords",
    "colours",
  ],
  "title": "DrawPixelAsset transaction asset for Canvas module",
  "type": "object",
}
`;
